### PR TITLE
Skip Private/Business/No Profile Pic, Min/Max posts

### PR DESCRIPTION
**Pull request by GabrieleCalarota**
#### This is used to skip users with certain condition
```python
session.set_skip_users(skip_private=True,
                       private_percentage=100,
                       skip_no_profile_pic=False,
                       no_profile_pic_percentage=100,
                       skip_business=False,
                       business_percentage=100,
                       skip_business_categories=[],
                       dont_skip_business_categories=[])
```
##### Skip private account
**This is done by default**
```python
session.set_skip_users(skip_private=True,
                       private_percentage=100)
```
Will skip users that have private account, even if are followed by running account.
You can set a percentage of skipping:
    _private_percentage_= 100 always skip private users
    _private_percentage_= 0 never skip private users (so set skip_private=False)

##### Skip users that don't have profile picture

```python
session.set_skip_users(skip_private=True,
                       skip_no_profile_pic=True,
                       no_profile_pic_percentage=100)
```
Will skip users that haven't uploaded yet a profile picture
You can set a percentage of skipping:
    _no_profile_pic_percentage_= 100 always skip users without profile picture
    _no_profile_pic_percentage_= 0 never skip users without profile picture (so set _skip_no_profile_pic_=False)

##### Skip users that have business account

```python
session.set_skip_users(skip_private=True,
                       skip_no_profile_pic=True,
		               skip_business=True,
		               business_percentage=100)
```
This will skip all users that have business account activated.
You can set a percentage of skipping:
    _business_percentage_= 100 always skip business users
    _business_percentage_= 0 never skip business users (so set _skip_business_=False)

**N.B.:** This _business_percentage_ parameter works only if no _skip_business_categories_ or _dont_skip_business_categories_ are provided!

###### Skip only users that have certain business account
```python
session.set_skip_users(skip_private=True,
                       skip_no_profile_pic=True,
		       skip_business=True,
		       skip_business_categories=['Creators & Celebrities'])
```
This will skip all business accounts that have category in given list
**N.B.** In _skip_business_categories_ you can add more than one category
###### Skip all business accounts, except from list given
```python
session.set_skip_users(skip_private=True,
                       skip_no_profile_pic=True,
		       skip_business=True,
		       dont_skip_business_categories=['Creators & Celebrities'])
```
This will skip all business accounts except the ones that have a category that matches one item in the list of _dont_skip_business_categories_
**N.B.** If both _dont_skip_business_categories_ and _skip_business_categories_, InstaPy will skip only business accounts in the list given from _skip_business_categories_.

### CODE REVIEW

**Comment by GabrieleCalarota** [comment]
#3048 
#2939

**Comment by GabrieleCalarota** [comment]
@timgrossmann I did what you requested in first place, what should i do for the bounty?

**Comment by timgrossmann** [comment]
@GabrieleCalarota Hey, thank you for the PR and taking part in this.

Unfortunately it does not work like this.
Your PR has to been merged and I have to close the issue in order for the bounty to be payed out.

As of now, your PR will be undergoing review in the next hours by @uluQulu @sionking @CharlesCCC me etc. 
Once it's approved we'll merge it and then you can claim the bounty.

I'll do an in depth review tomorrow, from what I see when glancing over it, I love the documentation!
However, the heroku thing is not possible... We need to keep it with Instagram directly and not rely on third party tooling.

**Comment by GabrieleCalarota** [comment]
@timgrossmann 
The code works I've tested it, but since I've made us of heroku app it is still a problem.

I don't like to use html tag to retrieve data (such as number of posts) cause maybe in two months they will update some html classes and it's going to be a new issue.

Using graphql (they change them less often) is a good idea, but we must have user_id instead of username for each user. Also I'm currently not sure if there is a graphql for such business accounts information. I'll try to figure it out.

**Comment by GabrieleCalarota** [comment]
@timgrossmann 
updated removing heroku, using browser.execute_script()

**Comment by sionking** [comment]
We do miss the option to just interact with some private users/  some business accounts.

**Comment by GabrieleCalarota** [comment]
For some business accounts you can just put skip_business=True and dont_skip_business_categories=[] list of the category you want to interact with.

If you meant to skip only a percentage of private account and business accounts I'll update that

**Comment by GabrieleCalarota** [comment]
@sionking updated

**Comment by sionking** [comment]
So I think this PR is more accurate to the ppl need I will close mine ( it was not for bounty anyway).
*Just next time if there going to be multiple ppl working on the same thing write it in the bounty.
* @timgrossmann  we still missing the validate_username function getting too big.
I think we shall create  a funtion that calls this function.

```
def validate_user_call(self)
   #call the validate_username()
                validation, details, params = validate_username(self.browser,
                                                        acc_to_follow,
                                                        self.username,
                                                        self.ignore_users,
                                                        self.blacklist,
                                                        self.potency_ratio,
                                                        self.delimit_by_numbers,
                                                        self.max_followers,
                                                        self.max_following,
                                                        self.min_followers,
                                                        self.min_following,
                                                        self.min_posts,
                                                        self.max_posts,
                                                        self.skip_private,
                                                        self.skip_private_percentage,
                                                        self.skip_no_profile_pic,
                                                        self.skip_no_profile_pic_percentage,
                                                        self.skip_business,
                                                        self.skip_business_percentage,
                                                        self.skip_business_categories,
                                                        self.dont_skip_business_categories,
                                                        self.logger)
   return validation, details, params 
```

**Comment by GabrieleCalarota** [comment]
We can do that, or even passing self as a parameter

**Comment by timgrossmann** [comment]
@sionking @GabrieleCalarota I'd prefer the separate method rather than simply passing `self` to the actual method.

**Comment by GabrieleCalarota** [comment]
Alright so I'll just stick with @sionking idea

File: instapy/instapy.py [hunk 1]
```diff
@@ -1123,28 +1127,7 @@ def like_by_locations(self,
...
-                                                                self.skip_private,
-                                                                self.skip_private_percentage,
-                                                                self.skip_no_profile_pic,
-                                                                self.skip_no_profile_pic_percentage,
-                                                                self.skip_business,
-                                                                self.skip_business_percentage,
-                                                                self.skip_business_categories,
-                                                                self.dont_skip_business_categories,
-                                                                self.logger)
+
```
Comment (timgrossmann) [inline instapy/instapy.py:1131]: Seems like you forgot it here

**Comment by timgrossmann** [comment]
@GabrieleCalarota Also, make sure to resolve the CI faults `F821 undefined name 'validation'`

**Comment by GabrieleCalarota** [comment]
@timgrossmann where was that fault from?

File: instapy/instapy.py (hunk 1 above)
Comment (GabrieleCalarota) [reply instapy/instapy.py:1131]: I've checked that twice, how could this happened in the first place.. anyway, fixed now ;)

**Comment by timgrossmann** [comment]
@GabrieleCalarota Thanks for the updates.
I'll test it tomorrow.
Would be great more people could do some testing for this feature.
@CharlesCCC @sionking @uluQulu etc..

Thanks guys 😉

**Comment by sionking** [comment]
I think we don't need **skip_business** **skip_private** as we use percentage private_percentage:
(0 == skip all, 100=never skip)

**Comment by GabrieleCalarota** [comment]
Yeah but percentage is not mandatory and also I think it is more easy to
understand skip_business=True rather than business_percentage=100.
Am I right ?

**Comment by sionking** [comment]
You also missing some details.
What if you don't skip private account ? will we go to "get_links_for_profile" ?
Maybe some xpath have changed now ?

Also checking private in tags and locations is unnecessary.

**Comment by GabrieleCalarota** [comment]
Also the code as it is right now, by default skip_business is False and if
you only set business_percentage=100 it won't work cause the first if
checks the skip_business Boolean parameter.

Also percentage in business is only used when no categories are provided,
as I written down in the documentation.

**Comment by GabrieleCalarota** [comment]
@sionking
Actually I didn't change anything, but I mean skip_private is by default
True, cause unless you are following that account or you requested to and
being accepted, you are not going to see anything cause there is also a
double check to see if the user is private.

This is after the validate username.

**Comment by sionking** [comment]
BUG:
You will check first for percentage in private and will not skip.
You will later reroll the percentage for following ignoring the roll for private.

I avoided that be returning values from validate_user.
Please have a look in interact_by_user at my PR.

File: instapy/instapy.py [hunk 2]
```diff
@@ -1926,7 +1927,7 @@ def interact_by_users(self,
             # static conditions
             not_dont_include = username not in self.dont_include
             follow_restricted = follow_restriction("read", username,
-                                                    self.follow_times, self.logger)
+                                                   self.follow_times, self.logger)
             counter = 0
             while True:
                 following = (random.randint(0, 100) <= self.follow_percentage and
```
Comment (sionking) [inline instapy/instapy.py:1933]: BUG: it will reroll the following check not according to private check

**Comment by GabrieleCalarota** [comment]
@sionking what's the point in enroll only once?
Also, even if your idea is good, I think this could be another feature non relevant to the pull request.. am I right ?

**Comment by GabrieleCalarota** [comment]
Because I'm sure this is not the only point where we enroll more than once a random chance ... Or are you just referring to the feature that if you don't want to skip a private user, you do want to follow him? @sionking

**Comment by GabrieleCalarota** [comment]
I can see your point now that I've looked at your PR, but I didn't get it when the issue and the request for feature was asked. I'm not saying that this is not related to this pull request, but maybe it wasn't specified. (Also because the percentage wasn't specified in the first place), so basically when the pull request was created you were either or skipping always or never skip. @sionking

I can add it, now.. but it will be add another parameter to all the validate_username.
Also I think that if you don't want to skip a users and the enroll of percentage passed, so you basically want that users to be followed. Am I right? So there's no point in passing that value, so if it's a private users and you validate him, you are going to follow him...

**Review by sionking — COMMENTED** [review]
> so if it's a private users and you validate him, you are going to follow him...

No that is not correct.
In real smart application where you put all your targets into poll I would agree with you, but now you will not control the amount of private users.
it can be 500 private and 100 non. Since you don't limit the it.
It will be still unpredictable but you can control it to lower and higher values.
Another thing is that you keep on validation the user (image-reco, and other stuff) since you think you are going to follow him, but you will not eventually.

**Comment by sionking** [comment]
> I can add it, now.. but it will be add another parameter to all the validate_username.

It should be another parameter **returned** from that function.

**Comment by GabrieleCalarota** [comment]
> > I can add it, now.. but it will be add another parameter to all the validate_username.
> 
> It should be another parameter **returned** from that function.

Yes I meant to all the validate_username calls. Even if it's only used in that particular case only for the skipping private feature.

**Comment by sionking** [comment]
> Yes I meant to all the validate_username calls. Even if it's only used in that particular case only for the skipping private feature.

Lets return a parameter that contain more than one item... it will be used in the feature for lets say "dont follow if number of posts but do likes" **all** parameters calculated in validate shall be in the caller function.

**Comment by GabrieleCalarota** [comment]
@timgrossmann when this will be merged?

**Comment by timgrossmann** [comment]
@GabrieleCalarota Sorry, haven't had time to test it, yet...
If we can find 2 more people to test it we can merge it.
Definitely will look through the code later today.

@sionking Can you approve this PR?

Could some of you guys do some testing, please? 
@fgisslen @jeremycjang @tompicca @lucanello @meinert @cormo1990 @aku6640 @zsvendo @the-unknown @redsector72 @bianpratama @Tachenz @aleeepp @johnvfitness @marcomokastyle @torbenberger @ArdySan2012 @Mehran @GabrieleCalarota @dbmathis @HermanoCrespo @loopypanda @Nocturnal-2 @darokif @Edhim @MMorte @sevilyilmaz @DimIsaev @SkylimiTY @ortegafernando @FranklinClinton @Wisdomlocker @riccardorighettiphoto @DanielSemmelrock @underclass @samille @clevebounce
(Thanks @uluQulu for the list 😛 )

Sorry for the delay.

Thank you
Best
Tim

**Comment by Mehran** [comment]
Tested Work good for me, Thanks @GabrieleCalarota @timgrossmann

**Comment by timgrossmann** [comment]
@GabrieleCalarota Looks good, let's wait for one more approval and we'll merge this. 👍

**Comment by Mehran** [comment]
@timgrossmann also Can add skipped if is Verified.

**Comment by timgrossmann** [comment]
@GabrieleCalarota Alright, we'll merge this now.

Let's see if there are any problems that will be uncovered once people use the feature in larger numbers.
Please make sure to fix them.
If there are no issues/all issues fixed, I will approve the claim in 24h.

I'll close the issues now so you can head over to https://www.bountysource.com/teams/instapy/issues?tracker_ids=45792278 and claim the bounties for the issues.

**Comment by timgrossmann** [comment]
fixes #2939 #980 #1006

**Comment by GabrieleCalarota** [comment]
@timgrossmann alright, thanks 😁

**Comment by timgrossmann** [comment]
@GabrieleCalarota Well, thank you!

**Comment by GabrieleCalarota** [comment]
@thefrankbot yes, if you want them not to be skipped just turn _skip_private_=False or _private_percentage_ to a value below 100 if you want to skip only some of them

**Comment by sionking** [comment]
@GabrieleCalarota So no solution for private percentage ?

**Comment by visualheroes** [comment]
Great work, thank you! Any idea where to find a list of business categories?

**Comment by sionking** [comment]
What about private user that we do follow ? get_images will fail I think ?

**Comment by GabrieleCalarota** [comment]
For private percentage I think I'll do a new PR next week @sionking 
Private users that we do follow I think it won't fail unless there are other checks after the validate usernames

**Comment by sionking** [comment]
Merge issue:
```
                    raise TimeoutException("Retried {} times to GET '{}' webpage "
                                           "but failed out of a timeout!\n\t{}".format(total_timeouts,
                                                                                       str(link).encode("utf-8"),
                                                                                       str(exc).encode("utf-8")))
```

**Comment by Edhim** [comment]
Hello, how to avoid skipping private profiles ?

**Comment by Abdulsamipy** [comment]
hi guys, so i was testing the skip_user_no_profile and it's not working any clue?

**Comment by Abdulsamipy** [comment]
hi guys, so i was testing  skip_no_profile_pic and instapy still following them any clue?
`    session.set_skip_users(skip_private=False,
                       private_percentage=100,
                       skip_no_profile_pic=True,                    
                       skip_business=False,
                       skip_non_business=False,
                       business_percentage=100)
`


image

**Pull request merged**